### DISCUSSION TITLE
[featurerequest]: [pilot] should provide mechanics to set AppStoreInfo.plist via -assetDescription option to iTMSTransporter

### DISCUSSION THREAD

**Discussion by cguentherTUChemnitz** [fastlane/fastlane#16993] [Ideas]
Question Checklist
-----
- [x]  Updated fastlane to the latest version
- [x]  I read the Contribution Guidelines
- [x]  I read docs.fastlane.tools
- [x]  I searched for existing GitHub issues

Motivation:
--------
Three tickets were aleady discussing this point, but where closed in front of a solution. So please see this ticket as kind of summarizing up and reopening with a specific goal and solution in mind.
https://github.com/fastlane/fastlane/issues/16131
https://github.com/fastlane/fastlane/issues/14349
https://github.com/fastlane/fastlane/issues/14256

We are using heavily gitlab-ci jobs for build, test and deploy jobs. Regarding to restrictions involved by XCode and in general mac, we are not able to scale mac job runner in the same way, as docker job runner. So our goal is to run only steps on the mac-runners where XCode is really enforced and bring every other step to docker based runners. This is especially important for long running jobs, like fastlane pilot app-store deploy, where the busy wait to see the change on the store entry can last >30min on some days.

**Main Goal: deploy under linux**

1. build and sign on mac (generating store.ipa and AppStoreInfo.plist if necessary)
 1.1 AppStoreInfo.plist can easily be generated after build running also on mac runner with `xcrun swinfo -f store.ipa -o AppStoreInfo.plist -prettyprint true --plistFormat binary
`
2. deploy on linux (deploying store.ipa)

Why is the new interface necessary?
------------
https://github.com/fastlane/fastlane/issues/14349
The current version of the itms >= 2 does need the AppStoreInfo.plist. This is generated automatically and on-the-fly on macs with installed xcode, but running itms under linux does need the previously created AppStoreInfo.plist. The problem is, the fastlane pilot interface do not provide an interface to add the `AppStoreInfo.plist` via `AppStoreInfo.plist` to `iTMSTransporter`.

<details><summary>fastlane and iTMSTransporter log</summary>

```
[06:47:46]: ----------------------------------
[06:47:46]: --- Step: upload_to_testflight ---
[06:47:46]: ----------------------------------
[06:47:46]: Login to App Store Connect (<censored>)
[06:47:51]: Login successful
[06:47:54]: Ready to upload new build to TestFlight (App: <censored>)...
[06:47:54]: Going to upload updated app to App Store Connect
[06:47:54]: This might take a few minutes. Please don't interrupt the script.
[06:48:06]: [Transporter Error Output]: Unable to perform software analysis on Linux. Export an AppStoreInfo.plist from Xcode, and use the -assetDescription option.
[06:48:06]: [Transporter Error Output]: Error while processing package 1461749712.itmsp
[06:48:06]: Transporter transfer failed.
[06:48:06]: 
[06:48:06]: Unable to perform software analysis on Linux. Export an AppStoreInfo.plist from Xcode, and use the -assetDescription option.
Error while processing package 1461749712.itmsp
[06:48:06]: [iTMSTransporter] [2020-08-05 06:48:06 CEST] <main> DEBUG: Handling asset: filename=bc9800267d627e27b8adeb1deb50d18f9b37df44bd86c09bf3fe9a452316eef7.ipa,uti=com.apple.ipa,role=source,describe=true

[06:48:06]: [iTMSTransporter] [2020-08-05 06:48:06 CEST] <main> DEBUG: Using container type tool name and version: iOS App Describer:1.26

[06:48:06]: [iTMSTransporter] [2020-08-05 06:48:06 CEST] <main> DEBUG: Using path: /tmp/d20200805-1-fe336a/1461749712.itmsp/bc9800267d627e27b8adeb1deb50d18f9b37df44bd86c09bf3fe9a452316eef7.ipa

[06:48:06]: [iTMSTransporter] [2020-08-05 06:48:06 CEST] <main> ERROR: Unable to perform software analysis on Linux. Export an AppStoreInfo.plist from Xcode, and use the -assetDescription option.

[06:48:06]: [iTMSTransporter] [2020-08-05 06:48:06 CEST] <main> ERROR: Error while processing package 1461749712.itmsp

[06:48:06]: [iTMSTransporter] 

[06:48:06]: [iTMSTransporter] 

[06:48:06]: [iTMSTransporter] 

[06:48:06]: [iTMSTransporter] Package Summary:

[06:48:06]: [iTMSTransporter]  

[06:48:06]: [iTMSTransporter] 1 package(s) were not uploaded because they had problems:

[06:48:06]: [iTMSTransporter] 	/tmp/d20200805-1-fe336a/1461749712.itmsp - Error Messages:

[06:48:06]: [iTMSTransporter] 		Unable to perform software analysis on Linux. Export an AppStoreInfo.plist from Xcode, and use the -assetDescription option.

[06:48:06]: [iTMSTransporter] 		Error while processing package 1461749712.itmsp

[06:48:06]: [iTMSTransporter] [2020-08-05 06:48:06 CEST] <main> DBG-X: Returning 1

[06:48:06]: iTunes Transporter output above ^
[06:48:06]: Unable to perform software analysis on Linux. Export an AppStoreInfo.plist from Xcode, and use the -assetDescription option.
Error while processing package 1461749712.itmsp
\nError while processing package 1461749712.itmsp to perform software analysis on Linux. Export an AppStoreInfo.plist from Xcode, and use the -assetDescription option.
The call to the iTMSTransporter completed with a non-zero exit status: 1. This indicates a failure.
[06:48:06]: Could not download/upload from App Store Connect! It's probably related to your password or your internet connection.
```

</details>

This docker image is used for execution
-----------
Going through this ticket https://github.com/fastlane/fastlane/issues/14256 i was able to create a docker image containing fastlane with the linux build of apple's itms.

For me it is this one:
<details><summary>Dockerfile</summary>

```Dockerfile
FROM ruby:2.7.1-buster

ARG USERNAME=build-bot

ENV DEBIAN_FRONTEND noninteractive

RUN apt-get update && apt-get -y install --auto-remove --purge debsums tzdata locales curl ca-certificates git unzip
RUN gem install cocoapods -v 1.9.3
RUN gem install fastlane

## configure locale
RUN apt-get update \
    && DEBIAN_FRONTEND=noninteractive apt-get install -y locales \
    && sed -i -e 's/# en_US.UTF-8 UTF-8/en_US.UTF-8 UTF-8/' /etc/locale.gen \
    && dpkg-reconfigure --frontend=noninteractive locales \
    && update-locale LANG=en_US.UTF-8
ENV LANG en_US.UTF-8
ENV LC_ALL en_US.UTF-8

## configure timezone
ENV TZ=Europe/Berlin
RUN ln -snf /usr/share/zoneinfo/$TZ /etc/localtime && echo $TZ > /etc/timezone

# You cannot run CocoaPods as root
RUN adduser "$USERNAME"
USER "$USERNAME"

# for new apple itms binary: download the install script
# https://appstoreconnect.apple.com/WebObjects/iTunesConnect.woa/ra/resources/download/Transporter__Linux/bin/
# accept licences
# copy /usr/local/itms folder to docker image build folder
ADD itms /usr/local/itms
ENV FASTLANE_ITUNES_TRANSPORTER_PATH /usr/local/itms/
ENV PATH $PATH:/usr/local/itms/bin

RUN sudo apt update && sudo apt install -y openjdk-11-jre

# add xar archive support
RUN cd /tmp \
    && git clone https://github.com/zentralopensource/xar \
    && cd xar/xar \
    && ./autogen.sh --noconfigure \
    && ./configure \
    && make -j 8 \
    && sudo make install \
    && rm -rf /tmp/xar

# use shell script does fix incompatbile java versions in calls between fastlane and itms
ENV FASTLANE_ITUNES_TRANSPORTER_USE_SHELL_SCRIPT 1
ENV DELIVER_ITMSTRANSPORTER_ADDITIONAL_UPLOAD_PARAMETERS "-t DAV #"
```

</details>

<details><summary>✅ fastlane environment ✅</summary>

### Stack

| Key                         | Value                                                    |
| --------------------------- | -------------------------------------------------------- |
| OS                          | debian_based 10.4                                        |
| Ruby                        | 2.7.1                                                    |
| Bundler?                    | false                                                    |
| Git                         | git version 2.20.1                                       |
| Installation Source         | /usr/local/bundle/bin/fastlane                           |
| Host                        | Debian GNU/Linux 10 debian_based 10.4 (5.6.19-2-MANJARO) |
| Ruby Lib Dir                | /usr/local/lib                                           |
| OpenSSL Version             | OpenSSL 1.1.1d  10 Sep 2019                              |
| Is contained                | false                                                    |
| Is homebrew                 | false                                                    |
| Is installed via Fabric.app | false                                                    |

### System Locale

| Variable | Value       |   |
| -------- | ----------- | - |
| LANG     | en_US.UTF-8 | ✅ |
| LC_ALL   | en_US.UTF-8 | ✅ |
| LANGUAGE |             |   |

### fastlane files:

**No Fastfile found**

**No Appfile found**

### fastlane gems

| Gem      | Version | Update-Status |
| -------- | ------- | ------------- |
| fastlane | 2.154.0 | ✅ Up-To-Date  |

### Loaded fastlane plugins:

**No plugins Loaded**

<details><summary><b>Loaded gems</b></summary>

| Gem                       | Version      |
| ------------------------- | ------------ |
| did_you_mean              | 1.4.0        |
| slack-notifier            | 2.3.2        |
| atomos                    | 0.1.3        |
| CFPropertyList            | 3.0.2        |
| claide                    | 1.0.3        |
| colored2                  | 3.1.2        |
| nanaimo                   | 0.3.0        |
| xcodeproj                 | 1.17.1       |
| rouge                     | 2.0.7        |
| xcpretty                  | 0.3.0        |
| terminal-notifier         | 2.0.0        |
| unicode-display_width     | 1.7.0        |
| terminal-table            | 1.8.0        |
| plist                     | 3.5.0        |
| public_suffix             | 4.0.5        |
| addressable               | 2.7.0        |
| multipart-post            | 2.0.0        |
| word_wrap                 | 1.0.0        |
| tty-screen                | 0.8.1        |
| tty-cursor                | 0.7.1        |
| tty-spinner               | 0.9.3        |
| babosa                    | 1.0.3        |
| colored                   | 1.2          |
| highline                  | 1.7.10       |
| commander-fastlane        | 4.4.6        |
| excon                     | 0.76.0       |
| faraday                   | 1.0.1        |
| unf_ext                   | 0.0.7.7      |
| unf                       | 0.1.4        |
| domain_name               | 0.5.20190701 |
| http-cookie               | 1.0.3        |
| faraday-cookie_jar        | 0.0.6        |
| faraday_middleware        | 1.0.0        |
| fastimage                 | 2.2.0        |
| gh_inspector              | 1.1.3        |
| json                      | 2.3.0        |
| mini_magick               | 4.10.1       |
| rubyzip                   | 2.3.0        |
| security                  | 0.1.3        |
| xcpretty-travis-formatter | 1.0.0        |
| dotenv                    | 2.7.6        |
| bundler                   | 2.1.4        |
| naturally                 | 2.2.0        |
| simctl                    | 1.6.8        |
| jwt                       | 2.2.1        |
| uber                      | 0.1.0        |
| declarative               | 0.0.20       |
| declarative-option        | 0.1.0        |
| representable             | 3.0.4        |
| retriable                 | 3.1.2        |
| mini_mime                 | 1.0.2        |
| multi_json                | 1.15.0       |
| signet                    | 0.14.0       |
| memoist                   | 0.16.2       |
| os                        | 1.1.1        |
| googleauth                | 0.13.1       |
| httpclient                | 2.8.3        |
| google-api-client         | 0.38.0       |
| google-cloud-env          | 1.3.3        |
| google-cloud-errors       | 1.0.1        |
| google-cloud-core         | 1.5.0        |
| rake                      | 13.0.1       |
| digest-crc                | 0.6.1        |
| google-cloud-storage      | 1.27.0       |
| emoji_regex               | 3.0.0        |
| jmespath                  | 1.4.0        |
| aws-partitions            | 1.350.0      |
| aws-eventstream           | 1.1.0        |
| aws-sigv4                 | 1.2.1        |
| aws-sdk-core              | 3.104.3      |
| aws-sdk-kms               | 1.36.0       |
| aws-sdk-s3                | 1.75.0       |
| uri                       | 0.10.0       |
| forwardable               | 1.3.1        |
| logger                    | 1.4.2        |
| cgi                       | 0.1.0        |
| timeout                   | 0.1.0        |
| stringio                  | 0.1.0        |
| ipaddr                    | 1.2.2        |
| openssl                   | 2.1.2        |
| ostruct                   | 0.2.0        |
| strscan                   | 1.0.3        |
| date                      | 3.0.0        |
| delegate                  | 0.1.0        |
| fileutils                 | 1.4.1        |
| etc                       | 1.1.0        |
| io-console                | 0.5.6        |
| zlib                      | 1.1.0        |
| singleton                 | 0.1.0        |
| rexml                     | 3.2.3        |
| open3                     | 0.1.0        |
| yaml                      | 0.1.0        |
| psych                     | 3.1.0        |
| mutex_m                   | 0.1.0        |
| webrick                   | 1.6.0        |
</details>

*generated on:* **2020-08-05**
</details>

The itms linux build installer has to be downloaded beforhand, via https://appstoreconnect.apple.com/WebObjects/iTunesConnect.woa/ra/resources/download/Transporter__Linux/bin/. I did not have automated the extraction from the installer. So currently the Dockerfile expects the extracted itms folder to live next to the Dockerfile.

**Comment by ciekawy** [comment]
any progress on that?
